### PR TITLE
Parser: add recovery for missing fields after 'of' in union case

### PR DESCRIPTION
**Pull request by auduchinok**
Fixes parsing and subsequent analysis for unfinished union cases and exceptions like
```fsharp
type U =
    | A of

exception E of
```

Before:
Screenshot 2020-12-15 at 00 21 02

After:
Screenshot 2020-12-15 at 00 20 42

### CODE REVIEW

**Comment by auduchinok** [comment]
There's some tests failure, but tests got broken again, so I can't debug it: https://github.com/dotnet/fsharp/pull/10708

**Comment by cartermp** [comment]
FCS tests pass, this one is an annoying timeout based one that seems to fail from time to time:

`FSharp.Core.UnitTests.Control.AsyncModule.AwaitWaitHandle.Timeout`

So restarting the PR or pushing a new change will probably have it resolve.

fsharpqa test failure looks like it's just a bad expectation, changing this:

https://github.com/dotnet/fsharp/blob/d76f5acb9a7fd526e51159489846a057bc4b1bd2/tests/fsharpqa/Source/Conformance/BasicTypeAndModuleDefinitions/UnionTypes/W_UnionCaseProduction01.fsx#L7

to this:

```diff
-  span="(10,12-11,1)"
+  span="(10,12-12,24)"
```

should fix it I think. The fsharpqa logs at least show that this was the range reported with your PR, which looks more accurate.

**Comment by cartermp** [comment]
grr, that should have done it. This is the failure log from CI:

```
+++ Conformance\BasicTypeAndModuleDefinitions\UnionTypes (W_UnionCaseProduction01.fsx) +++
--------------------------------------------------------
Results from hosted compiler
msg: Compiling
cmd: fsc   -a --test:ErrorRanges W_UnionCaseProduction01.fsx 
Exit code: 0
Error:     1
Microsoft (R) F# Compiler version 11.0.0.0 for F# 5.0
Copyright (c) Microsoft Corporation. All Rights Reserved.
(10,12-10,24): warning FS0042: This construct is deprecated: it is only for use in the F# library
(10,12-10,24): warning FS0042: This construct is deprecated: it is only for use in the F# library
--------------------------------------------------------
Microsoft (R) F# Compiler version 11.0.0.0 for F# 5.0
Copyright (c) Microsoft Corporation. All Rights Reserved.
(10,12-10,24): warning FS0042: This construct is deprecated: it is only for use in the F# library
(10,12-10,24): warning FS0042: This construct is deprecated: it is only for use in the F# library

*** The following necessary lines were never matched:
***	\(10,12-12,24\):.+warning FS0042:.+This construct is deprecated: it is only for use in the F# library


*** The following necessary lines were incorrectly matched:

Unexpected Compiler Output 
FAIL
```

Does that imply that the warning was given twice or something? I don't know how to read this (cc @KevinRansom @dsyme)

**Comment by auduchinok** [comment]
OK, let's see if it helps (though, it's not expected indeed).

File: tests/fsharpqa/Source/Conformance/BasicTypeAndModuleDefinitions/UnionTypes/W_UnionCaseProduction01.fsx [hunk 1]
```diff
@@ -4,7 +4,8 @@
 //  | id	-- nullary union case
 //  | id of type * ... * type	-- n-ary union case
 //  | id : sig-spec	-- n-ary union case
-//<Expects id="FS0042" span="(10,12-11,1)" status="warning">This construct is deprecated: it is only for use in the F# library</Expects>
+//<Expects id="FS0042" span="(10,12-12,24)" status="warning">This construct is deprecated: it is only for use in the F# library</Expects>
+//<Expects id="FS0042" span="(10,12-12,24)" status="warning">This construct is deprecated: it is only for use in the F# library</Expects>
```
Comment (cartermp) [inline tests/fsharpqa/Source/Conformance/BasicTypeAndModuleDefinitions/UnionTypes/W_UnionCaseProduction01.fsx:8]:
```suggestion
//<Expects id="FS0042" span="(11,12-11,24)" status="warning">This construct is deprecated: it is only for use in the F# library</Expects>
//<Expects id="FS0042" span="(11,12-11,24)" status="warning">This construct is deprecated: it is only for use in the F# library</Expects>
```

**Review by cartermp — COMMENTED** [review]
this should be the one that actually does it if I'm reading the log correctly.

Although it's strange that we would get two diagnostics.

**Review by cartermp — APPROVED** [review]

**Pull request merged**